### PR TITLE
Prep for 3.0 Release

### PR DESCRIPTION
**Pull request by pengale**
This Draft PR contains bitrot fixes, tests, and the first pass at an interface library for the dashboard charm.

### CODE REVIEW

**Comment by anthonydillon** [comment]
@jpmartinspt would you mind having a look over these changes, please?

File: k8s-charm-sidecar/CONTRIBUTING.md
Comment (jkfran) [inline k8s-charm-sidecar/CONTRIBUTING.md:17]:
```suggestion
along with Juju, and installed automatically.
```

File: k8s-charm-sidecar/metadata.yaml
Comment (jkfran) [inline k8s-charm-sidecar/metadata.yaml:1]: You have this copyright notice in multiple files under your name but on k8s-charm-sidecar/charmcraft.yaml for example is `# Copyright 2022 Canonical Ltd.`. Shouldn't be Canonical Ltd everywhere?

**Review by jkfran — APPROVED** [review]
Code wise it looks good to me, I added two small comments.

File: README.md
Comment (wallyworld) [inline README.md:53]: The strict microk8s snap appears to have been released to `1.25-strict/stable`.
I am not exactly sure of the deprecation schedule for the classic snap but I think we should be transitioning to strict. Juju 3 integrates with the strict version and we will drop support for microk8s classic at some point I expect. With strict, I am not sure that you still need to be in the microk8s group anymore, but that needs fact checking.

**Comment by anthonydillon** [comment]
@jkfran could you give this PR another review and merge, if it looks good, please.

**Comment by jkfran** [comment]
> @jkfran could you give this PR another review and merge, if it looks good, please.

There are many `TODO` in the changes I am not the PR is meant to be merged yet.

**Comment by barrettj12** [comment]
@jkfran we're planning to release Juju 3.0 before next week, we would like a stable release of the `juju-dashboard` to accompany that. AFAIK the current beta on Charmhub reflects the tip of this branch. I have tested the changes here and everything seems to work (aside from #17, but I already have a [patch for this](https://github.com/pengale/juju-dashboard-charm/compare/main...barrettj12:juju-dashboard-charm:ip-address) which I will merge in once this PR lands).

Yes, there are some TODOs, but most of them are in documentation and testing, and the few that are in the actual code don't seriously affect the charm's operation. AFAIK pengale (original author) is no longer at Canonical so won't be expected to make any more changes here. In any case what's on this branch is a big improvement over the current main, so I think we're best to just merge it.

I'm happy to continue working on TODOs / bug fixes going forward, but as I say, we (Juju team) would like something "releasable" for now.

**Comment by goulinkh** [comment]
@barrettj12 @wallyworld I've updated the instructions on how to deploy the juju dashboard on microk8s as the current version of the dashboard has some changes and is compatible with Juju 3.0 without any manual changes!

However the dashboard didn't work locally on my machine, I managed to depoy it to microk8s but the command:
```
$ juju dashboard
Dashboard for controller "micro-localhost" is enabled at:
  http://localhost:34503
Your login credential is:
  username: admin
  password: 1422254fd985b7e8b279f061cb36d2ac
```

Returns the HTTP page but when signing in the websocket server isn't accessible: `ws://localhost:34503/api`

**Comment by barrettj12** [comment]
Hi @goulinkh, I've reproduced your issue on microk8s. I'm also seeing that the websocket is not accessible to the dashboard frontend.

While this is worrying, I think it is an issue with the dashboard itself (on microk8s), rather than our charm. I tried with the current charm in canonical:main and got exactly the same issue (`WebSocket connection to 'ws://localhost:45583/api' failed`). So there appears to be no regression in this PR.

In fact, with the current version (rev 2) of `juju-dashboard-k8s`, it is even worse - the dashboard gives `nginx 404 not found`. So I maintain that what we have here is a significant improvement, and we should merge it. I will continue looking into this dashboard issue in the coming weeks.

**Comment by barrettj12** [comment]
There are issues on LXD too but again, these appear to be problems with the dashboard itself, **not** with this charm:

```
Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'controllerConfig')
    at index.js:345:60
    at model-poller.js:105:3
```

```
TypeError: Cannot read properties of undefined (reading 'listModels')
    at model-poller.js:120:54
```

I'll raise these issues on the Juju dashboard repo, meanwhile I suggest we merge the work we have here.

**Comment by goulinkh** [comment]
> I'll raise these issues on the Juju dashboard repo, meanwhile I suggest we merge the work we have here.

These issues are fixed in the stable version of the Juju dashboard, I will update the dashboard for LXD to the latest version and merge this PR :+1:

**Comment by anthonydillon** [comment]
@goulinkh labels.

**Pull request merged**
